pkg/controller/mpi_job_controller.go

```diff
@@ -628,15 +628,12 @@ func (c *MPIJobController) syncHandler(key string) error {
 				return err
 			}
 		}
-		if mpiJob.Spec.MPIImplementation == kubeflow.MPIImplementationIntel ||
-			mpiJob.Spec.MPIImplementation == kubeflow.MPIImplementationMPICH {
-			// The Intel and MPICH implementations require workers to communicate with the
-			// launcher through its hostname. For that, we create a Service which
-			// has the same name as the launcher's hostname.
-			_, err := c.getOrCreateService(mpiJob, newLauncherService(mpiJob))
-			if err != nil {
-				return fmt.Errorf("getting or creating Service to front launcher: %w", err)
-			}
+		// NEW: always create service for launcher for different implementations and compatible with running launcher as worker
+		// The Intel and MPICH implementations require workers to communicate with the
+		// launcher through its hostname. For that, we create a Service which
+		// has the same name as the launcher's hostname.
+		if _, err = c.getOrCreateService(mpiJob, newLauncherService(mpiJob)); err != nil {
```

Comment:
```suggestion
		if mpiJob.Spec.RunLauncherAsWorker || mpiJob.Spec.MPIImplementation == kubeflow.MPIImplementationIntel ||
			mpiJob.Spec.MPIImplementation == kubeflow.MPIImplementationMPICH {		
				if _, err = c.getOrCreateService(mpiJob, newLauncherService(mpiJob)); err != nil {
```

Why do we always need to create a Service for the Launcher?
IIUC, in the case of `runLauncherAsWorker=false`, creating service isn't needed.

Reply:
Done, I did want to keep it simple. It's OK this way.